plugins/genesys/fu-genesys-usbhub-dev-firmware.c

```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright (C) 2023 Adam.Chen <Adam.Chen@genesyslogic.com.tw>
+ *
+ * SPDX-License-Identifier: LGPL-2.1+
+ */
+
+#include "config.h"
+
+#include "fu-genesys-common.h"
+#include "fu-genesys-usbhub-dev-firmware.h"
+#include "fu-genesys-usbhub-firmware.h"
+#include "fu-genesys-usbhub-struct.h"
+
+struct _FuGenesysUsbhubDevFirmware {
+	FuFirmwareClass parent_instance;
+};
+
+G_DEFINE_TYPE(FuGenesysUsbhubDevFirmware, fu_genesys_usbhub_dev_firmware, FU_TYPE_FIRMWARE)
+
+static gboolean
+fu_genesys_usbhub_dev_firmware_check_magic(FuFirmware *firmware,
+					   GBytes *fw,
+					   gsize offset,
+					   GError **error)
+{
+	guint8 magic[GENESYS_USBHUB_FW_SIG_LEN] = {0x0};
+
+	if (!fu_memcpy_safe(magic,
+			    sizeof(magic),
+			    0, /* dst */
+			    g_bytes_get_data(fw, NULL),
+			    g_bytes_get_size(fw),
+			    offset + GENESYS_USBHUB_FW_SIG_OFFSET,
+			    GENESYS_USBHUB_FW_SIG_LEN,
+			    error)) {
+		g_prefix_error(error, "failed to read magic: ");
+		return FALSE;
+	}
+	if (memcmp(magic, GENESYS_USBHUB_FW_SIG_TEXT_DEV_BRIDGE, GENESYS_USBHUB_FW_SIG_LEN) != 0) {
+		g_set_error_literal(error,
+				    FWUPD_ERROR,
+				    FWUPD_ERROR_INVALID_FILE,
+				    "signature not supported");
+		return FALSE;
+	}
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_genesys_usbhub_dev_firmware_parse(FuFirmware *firmware,
+				     GBytes *fw,
+				     gsize offset,
+				     FwupdInstallFlags flags,
+				     GError **error)
+{
+	gsize bufsz = 0;
+	const guint8 *buf = g_bytes_get_data(fw, &bufsz);
+	guint32 code_size = 0;
+	g_autoptr(GBytes) blob = NULL;
+
+	fu_firmware_set_id(firmware, fu_genesys_fw_type_to_string(FU_GENESYS_FW_TYPE_DEV_BRIDGE));
+	fu_firmware_set_idx(firmware, FU_GENESYS_FW_TYPE_DEV_BRIDGE);
+	fu_firmware_set_alignment(firmware, FU_FIRMWARE_ALIGNMENT_1K);
+	fu_firmware_set_offset(firmware, offset);
+
+	bufsz -= offset;
+	buf += offset;
+
+	/* deduce code size */
+	if (!fu_genesys_usbhub_firmware_query_codesize(firmware, buf, bufsz, error)) {
+		g_prefix_error(error, "not valid for dev:");
+		return FALSE;
+	}
+	code_size = fu_firmware_get_size(firmware);
+
+	/* calculate checksum */
+	if ((flags & FWUPD_INSTALL_FLAG_IGNORE_CHECKSUM) == 0) {
+		if (!fu_genesys_usbhub_firmware_verify(buf, bufsz, code_size, error)) {
+			g_prefix_error(error, "not valid for dev:");
+			return FALSE;
+		}
+	}
+
+	blob = fu_bytes_new_offset(fw, offset, code_size, error);
+	if (blob == NULL) {
+		g_prefix_error(error, "not valid for dev:");
+		return FALSE;
+	}
+	fu_firmware_set_bytes(firmware, blob);
+
+	/* get firmware version */
+	if (!fu_genesys_usbhub_firmware_query_version(firmware, buf, bufsz, error)) {
+		g_prefix_error(error, "not valid for dev:");
```

Comment:
needs an extra space after the trailing :

Reply:
searched related using and done.